core/blockchain.go

```diff
@@ -659,65 +686,70 @@ func (self *BlockChain) LoadLastState(dryrun bool) error {
-	highestApparentHead := self.hc.CurrentHeader().Number.Uint64()
+	highestApparentHead := bc.hc.CurrentHeader().Number.Uint64()
 	aboveHighestApparentHead := highestApparentHead + 2048
 
-	if b := self.GetBlockByNumber(aboveHighestApparentHead); b != nil {
+	if b := bc.GetBlockByNumber(aboveHighestApparentHead); b != nil {
 		glog.V(logger.Warn).Errorf("Found block data beyond apparent head (head=%d, found=%d)", highestApparentHead, aboveHighestApparentHead)
 		return recoverOrReset()
 	}
 
 	// Check head block number congruent to hash.
-	if b := self.GetBlockByNumber(self.currentBlock.NumberU64()); b != nil && b.Header() != nil && b.Header().Hash() != self.currentBlock.Hash() {
-		glog.V(logger.Error).Errorf("Found head block number and hash mismatch: number=%d, hash=%x", self.currentBlock.NumberU64(), self.currentBlock.Hash())
+	if b := bc.GetBlockByNumber(bc.currentBlock.NumberU64()); b != nil && b.Header() != nil && b.Header().Hash() != bc.currentBlock.Hash() {
+		glog.V(logger.Error).Errorf("Found head block number and hash mismatch: number=%d, hash=%x", bc.currentBlock.NumberU64(), bc.currentBlock.Hash())
 		return recoverOrReset()
 	}
 
 	// Check head header number congruent to hash.
-	if h := self.hc.GetHeaderByNumber(self.hc.CurrentHeader().Number.Uint64()); h != nil && self.hc.GetHeader(h.Hash()) != h {
-		glog.V(logger.Error).Errorf("Found head header number and hash mismatch: number=%d, hash=%x", self.hc.CurrentHeader().Number.Uint64(), self.hc.CurrentHeader().Hash())
+	if h := bc.hc.GetHeaderByNumber(bc.hc.CurrentHeader().Number.Uint64()); h != nil && bc.hc.GetHeader(h.Hash()) != h {
+		glog.V(logger.Error).Errorf("Found head header number and hash mismatch: number=%d, hash=%x", bc.hc.CurrentHeader().Number.Uint64(), bc.hc.CurrentHeader().Hash())
 		return recoverOrReset()
 	}
 
 	// Case: current header below fast or full block.
 	//
-	highestCurrentBlockFastOrFull := self.currentBlock.Number()
-	if self.currentFastBlock.Number().Cmp(highestCurrentBlockFastOrFull) > 0 {
-		highestCurrentBlockFastOrFull = self.currentFastBlock.Number()
+	highestCurrentBlockFastOrFull := bc.currentBlock.Number()
+	if bc.currentFastBlock.Number().Cmp(highestCurrentBlockFastOrFull) > 0 {
+		highestCurrentBlockFastOrFull = bc.currentFastBlock.Number()
 	}
 	// If the current header is behind head full block OR fast block, we should reset to the height of last OK header.
-	if self.hc.CurrentHeader().Number.Cmp(highestCurrentBlockFastOrFull) < 0 {
+	if bc.hc.CurrentHeader().Number.Cmp(highestCurrentBlockFastOrFull) < 0 {
 		glog.V(logger.Error).Errorf("Found header height below block height, attempting reset with recovery...")
 		return recoverOrReset()
 	}
 
+	// If header is beyond genesis, but both current and current fast blocks are 0, maybe something has gone wrong,
+	// like a write error on previous shutdown.
+	if bc.hc.CurrentHeader().Number.Cmp(big.NewInt(0)) > 0 && highestCurrentBlockFastOrFull.Cmp(big.NewInt(0)) == 0 {
```

Comment:
Another sanity check for a possible recovery flag. I've had this happen to me once. If we ever do "LightChain"s this check will be invalid, but for now both fast and full syncs will satisfy this demand.